File: src/test/java/com/flowcommerce/core/domain/member/SellerApplyValidatorTest.java
package com.flowcommerce.core.domain.member;

import com.flowcommerce.core.domain.User;
import com.flowcommerce.core.storage.member.SellerEntity;
import com.flowcommerce.core.storage.member.SellerRepository;
import com.flowcommerce.core.support.error.CoreException;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;

@ExtendWith(MockitoExtension.class)
class SellerApplyValidatorTest {

    @Mock
    SellerRepository sellerRepository;

    @Mock
    AccountVerificationService accountVerificationService;

    @InjectMocks
    SellerApplyValidator sellerApplyValidator;

    @Test
    void 정상_판매자_신청_검증_성공() {
        // given
        User user = new User(1L);
        SellerApply sellerApply = SellerApplyFixture.createDefault();

        given(sellerRepository.findByMemberId(any())).willReturn(null);
        given(sellerRepository.existsBySellerName(any())).willReturn(false);
        given(sellerRepository.existsByBusinessNumber(any())).willReturn(false);
        given(accountVerificationService.verify(any(), any(), any())).willReturn(true);

        // when, then
        assertDoesNotThrow(() -> sellerApplyValidator.validateApply(user, sellerApply));
    }

    @Test
    void 이미_승인된_판매자일_경우_신청실패(){
        // given
        User user = new User(1L);
        SellerApply sellerApply = SellerApplyFixture.createDefault();
        SellerEntity seller = SellerEntityFixture.createWithStatus(SellerStatus.APPROVED);

        given(sellerRepository.findByMemberId(any())).willReturn(seller);

        // when, then
        assertThatThrownBy(() -> sellerApplyValidator.validateApply(user, sellerApply))
                .isInstanceOf(CoreException.class);
    }

    @Test
    void 판매자_신청대기중인_판매자일_경우_신청실패(){
        // given
        User user = new User(1L);
        SellerApply sellerApply = SellerApplyFixture.createDefault();
        SellerEntity seller = SellerEntityFixture.createWithStatus(SellerStatus.PENDING);

        given(sellerRepository.findByMemberId(any())).willReturn(seller);

        // when, then
        assertThatThrownBy(() -> sellerApplyValidator.validateApply(user, sellerApply))
                .isInstanceOf(CoreException.class);
    }

    @Test
    void 계좌정보가_일치하지_않을경우_신청실패() {
        // given
        User user = new User(1L);
        SellerApply sellerApply = SellerApplyFixture.createDefault();

        given(sellerRepository.findByMemberId(any())).willReturn(null);
        given(sellerRepository.existsBySellerName(any())).willReturn(false);
        given(sellerRepository.existsByBusinessNumber(any())).willReturn(false);
        given(accountVerificationService.verify(any(), any(), any())).willReturn(false);

        // when, then
        assertThatThrownBy(() -> sellerApplyValidator.validateApply(user, sellerApply))
                .isInstanceOf(CoreException.class);
    }

    @Test
    void 판매자명_중복일때_신청실패() {
        // given
        User user = new User(1L);
        SellerApply sellerApply = SellerApplyFixture.createDefault();

        given(sellerRepository.findByMemberId(any())).willReturn(null);
        given(sellerRepository.existsBySellerName(any())).willReturn(true);

        // when, then
        assertThatThrownBy(() -> sellerApplyValidator.validateApply(user, sellerApply))
                .isInstanceOf(CoreException.class);
    }

    @Test
    void 사업자번호_중복일때_신청실패() {
        // given
        User user = new User(1L);
        SellerApply sellerApply = SellerApplyFixture.createDefault();

        given(sellerRepository.findByMemberId(any())).willReturn(null);
        given(sellerRepository.existsBySellerName(any())).willReturn(false);
        given(sellerRepository.existsByBusinessNumber(any())).willReturn(true);

        // when, then
        assertThatThrownBy(() -> sellerApplyValidator.validateApply(user, sellerApply))
                .isInstanceOf(CoreException.class);
    }
}

File: src/test/kotlin/com/flowcommerce/core/domain/member/SellerApplyValidatorTest.kt
package com.flowcommerce.core.domain.member

import com.flowcommerce.core.domain.User
import com.flowcommerce.core.storage.member.SellerRepository
import com.flowcommerce.core.support.error.CoreException
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Assertions.assertDoesNotThrow
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.BDDMockito.given
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
class SellerApplyValidatorTest {

    @Mock
    lateinit var sellerRepository: SellerRepository

    @Mock
    lateinit var accountVerificationService: AccountVerificationService

    @InjectMocks
    lateinit var sellerApplyValidator: SellerApplyValidator

    @Test
    fun `정상 판매자 신청 검증 성공`() {
        // given
        val user = User(1L)
        val sellerApply = SellerApplyFixture.createDefault()

        given(sellerRepository.findByMemberId(user.id)).willReturn(null)
        given(sellerRepository.existsBySellerName(sellerApply.sellerName)).willReturn(false)
        given(sellerRepository.existsByBusinessNumber(sellerApply.businessNumber)).willReturn(false)
        given(accountVerificationService.verify(
            sellerApply.accountHolder,
            sellerApply.bankName,
            sellerApply.accountNumber
        )).willReturn(true)

        // when, then
        assertDoesNotThrow { sellerApplyValidator.validateApply(user, sellerApply) }
    }

    @Test
    fun `이미 승인된 판매자일 경우 신청실패`() {
        // given
        val user = User(1L)
        val sellerApply = SellerApplyFixture.createDefault()
        val seller = SellerEntityFixture.createWithStatus(SellerStatus.APPROVED)

        given(sellerRepository.findByMemberId(user.id)).willReturn(seller)

        // when, then
        assertThatThrownBy { sellerApplyValidator.validateApply(user, sellerApply) }
            .isInstanceOf(CoreException::class.java)
    }

    @Test
    fun `판매자 신청대기중인 판매자일 경우 신청실패`() {
        // given
        val user = User(1L)
        val sellerApply = SellerApplyFixture.createDefault()
        val seller = SellerEntityFixture.createWithStatus(SellerStatus.PENDING)

        given(sellerRepository.findByMemberId(user.id)).willReturn(seller)

        // when, then
        assertThatThrownBy { sellerApplyValidator.validateApply(user, sellerApply) }
            .isInstanceOf(CoreException::class.java)
    }

    @Test
    fun `계좌정보가 일치하지 않을경우 신청실패`() {
        // given
        val user = User(1L)
        val sellerApply = SellerApplyFixture.createDefault()

        given(sellerRepository.findByMemberId(user.id)).willReturn(null)
        given(sellerRepository.existsBySellerName(sellerApply.sellerName)).willReturn(false)
        given(sellerRepository.existsByBusinessNumber(sellerApply.businessNumber)).willReturn(false)
        given(accountVerificationService.verify(
            sellerApply.accountHolder,
            sellerApply.bankName,
            sellerApply.accountNumber
        )).willReturn(false)

        // when, then
        assertThatThrownBy { sellerApplyValidator.validateApply(user, sellerApply) }
            .isInstanceOf(CoreException::class.java)
    }

    @Test
    fun `판매자명 중복일때 신청실패`() {
        // given
        val user = User(1L)
        val sellerApply = SellerApplyFixture.createDefault()

        given(sellerRepository.findByMemberId(user.id)).willReturn(null)
        given(sellerRepository.existsBySellerName(sellerApply.sellerName)).willReturn(true)

        // when, then
        assertThatThrownBy { sellerApplyValidator.validateApply(user, sellerApply) }
            .isInstanceOf(CoreException::class.java)
    }

    @Test
    fun `사업자번호 중복일때 신청실패`() {
        // given
        val user = User(1L)
        val sellerApply = SellerApplyFixture.createDefault()

        given(sellerRepository.findByMemberId(user.id)).willReturn(null)
        given(sellerRepository.existsBySellerName(sellerApply.sellerName)).willReturn(false)
        given(sellerRepository.existsByBusinessNumber(sellerApply.businessNumber)).willReturn(true)

        // when, then
        assertThatThrownBy { sellerApplyValidator.validateApply(user, sellerApply) }
            .isInstanceOf(CoreException::class.java)
    }
}
